File: backend/internal/services/contracts/set_type.go
package contracts

import (
	"slices"
	"strings"
)

// ──────────────────────────────────────────────
// Bill roles (show_artists.set_type)
// ──────────────────────────────────────────────

// The curated vocabulary for show_artists.set_type.
//
// set_type answers "what slot did this act play on this bill", which is a
// CURATED fact: somebody has to know it. It is therefore never inferred from
// list order. Only SetTypeHeadliner has ever been inferable (from position or
// the legacy is_headliner flag), and it is the only inference this package
// still sanctions.
//
// SetTypePerformer is the neutral default: "on the bill, slot unknown". It
// must stay semantically empty, because it is what every uncurated row holds.
// Writing a specific role (opener, direct support, ...) as a default would
// publish a guess as a fact, which is exactly the defect this vocabulary
// replaces.
const (
	SetTypeHeadliner     = "headliner"
	SetTypeDirectSupport = "direct_support"
	SetTypeOpener        = "opener"
	SetTypeDJ            = "dj"
	SetTypeSpecialGuest  = "special_guest"
	SetTypePerformer     = "performer"
)

// SetTypeDefault is the value written when nothing curated is known about an
// act's slot.
//
// Every writer in this codebase supplies it rather than leaving the column to
// chance, and the PSY-1673 migration normalizes the rows that predate that
// rule, so readers may treat set_type as always-present. Note the SCHEMA does
// not enforce it: the column is `VARCHAR(50) DEFAULT 'performer'` with no NOT
// NULL and no CHECK, so this is a convention the code keeps, not a guarantee
// the database makes. Adding those constraints would make it one.
const SetTypeDefault = SetTypePerformer

// setTypeVocabulary is ordered top-of-bill first, then by descending
// specificity, so the API docs, the error message, and the form selector all
// present the same order.
var setTypeVocabulary = []string{
	SetTypeHeadliner,
	SetTypeDirectSupport,
	SetTypeOpener,
	SetTypeSpecialGuest,
	SetTypeDJ,
	SetTypePerformer,
}

// SetTypeVocabulary returns the accepted set_type values in presentation
// order. Returns a copy so callers cannot mutate the vocabulary.
func SetTypeVocabulary() []string {
	return slices.Clone(setTypeVocabulary)
}

// SetTypeVocabularyCSV renders the vocabulary as a comma-separated list, for
// OpenAPI enum tags and validation messages.
func SetTypeVocabularyCSV() string {
	return strings.Join(setTypeVocabulary, ",")
}

// IsValidSetType reports whether value is exactly one of the curated set_type
// values. STRICT on purpose: this is the API contract check, so a client that
// sends "Headliner" or "support" gets a 422 naming the field rather than a
// silently coerced role. Lenient coercion of third-party vocabulary is
// NormalizeSetType's job, and only ingest may use it.
func IsValidSetType(value string) bool {
	return slices.Contains(setTypeVocabulary, value)
}

// NormalizeSetType maps a set_type as stated by an outside source (a venue
// calendar, an AI extraction of a flyer) onto the curated vocabulary. It is
// the INGEST-side counterpart to IsValidSetType and must not be used to
// validate API input.
//
// Returns "" for anything it cannot map with confidence, which callers must
// read as "the source said nothing about this slot" and answer with
// SetTypeDefault. Guessing here would reintroduce the very defect this
// vocabulary exists to remove: a role label nobody actually asserted.
//
// Mapping table, and why each row is honest:
//
//	headliner       -> headliner       identical term
//	direct_support  -> direct_support  identical term
//	support         -> direct_support  the extraction prompt emits "support"
//	                                   ONLY for an act the source billed under
//	                                   "w/" or "with", which is the direct
//	                                   support slot by definition
//	opener          -> opener          the source said "opener"; a stated
//	                                   opening slot is a real curated fact
//	special_guest   -> special_guest   identical term
//	dj              -> dj              now a first-class slot, no longer
//	                                   flattened into performer
//	performer       -> performer       identical term
//	host / mc       -> ""              hosting is a real role the vocabulary
//	                                   does not model; "performer" would be a
//	                                   lossy guess, so the caller defaults
//	anything else   -> ""              unknown means unknown
func NormalizeSetType(value string) string {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case SetTypeHeadliner:
		return SetTypeHeadliner
	case SetTypeDirectSupport, "support":
		return SetTypeDirectSupport
	case SetTypeOpener:
		return SetTypeOpener
	case SetTypeSpecialGuest:
		return SetTypeSpecialGuest
	case SetTypeDJ:
		return SetTypeDJ
	case SetTypePerformer:
		return SetTypePerformer
	default:
		return ""
	}
}

File: cli/src/lib/setType.ts
/**
 * The curated bill roles the API accepts on `show_artists.set_type`.
 *
 * A client-side mirror of the enum published on the shared show `Artist`
 * schema, in the backend's presentation order (top of bill first). The backend
 * is the enforcement point and answers an out-of-vocabulary value with 422;
 * validating here only buys the operator a message that names the vocabulary
 * before a request goes out.
 */
export const SET_TYPE_VOCABULARY = [
  "headliner",
  "direct_support",
  "opener",
  "special_guest",
  "dj",
  "performer",
] as const;

export type SetType = (typeof SET_TYPE_VOCABULARY)[number];

/**
 * The neutral role: "on the bill, slot unknown".
 *
 * It is what every uncurated row holds, so it states nothing about a slot and
 * must never be written back as though somebody had curated it. Absent and
 * `performer` are the two spellings of the same silence.
 */
export const SET_TYPE_UNCURATED: SetType = "performer";

/** Whether `value` is exactly one of the accepted roles. */
export function isValidSetType(value: string): value is SetType {
  return (SET_TYPE_VOCABULARY as readonly string[]).includes(value);
}

/** The vocabulary as a comma-separated list, for validation messages. */
export const SET_TYPE_VOCABULARY_CSV = SET_TYPE_VOCABULARY.join(", ");

/**
 * Whether a stored value says anything at all about an act's slot.
 *
 * THE definition of silence, so the two questions built on it cannot disagree:
 * what gets sent back, and what gets warned about. Absent, blank and
 * `performer` are its three spellings.
 */
function statesASlot(value: string | null | undefined): value is string {
  if (value == null) return false;
  const trimmed = value.trim();
  return trimmed !== "" && trimmed !== SET_TYPE_UNCURATED;
}

/**
 * The role to send back for an act read off a show response, or `undefined`
 * when the stored value curates nothing.
 *
 * Two values resolve to `undefined` and they mean different things:
 *
 * - Silence, per {@link statesASlot}. Only an ABSENT key says "slot unknown" on
 *   the way back in, so the caller must OMIT the field rather than send
 *   `performer` or `null`.
 * - A value outside the vocabulary is a row the API would refuse. Rewriting
 *   the bill is the only way this CLI can add or drop an act, so the choice is
 *   between dropping that one unreadable value and failing the whole edit;
 *   callers drop it and say so out loud. The column carries no CHECK
 *   constraint, so such a row is reachable.
 *
 * The value is judged EXACTLY as stored, untrimmed, which is the same verdict
 * the API gives a role stated on the way in. Only emptiness is forgiving.
 */
export function curatedSetType(
  value: string | null | undefined,
): SetType | undefined {
  if (!statesASlot(value)) return undefined;
  return isValidSetType(value) ? value : undefined;
}

/**
 * Whether a stored value is one this CLI cannot round-trip: it states a slot,
 * but not one the API would accept back.
 *
 * Derived from {@link curatedSetType} rather than restating its test, so the
 * value that gets dropped is exactly the value that gets warned about.
 */
export function isUnroundtrippableSetType(
  value: string | null | undefined,
): boolean {
  return statesASlot(value) && curatedSetType(value) === undefined;
}
